src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs

```diff
@@ -8845,6 +8867,186 @@ public void Version()
+
+            VerifyPEMetadata(File.OpenRead(exe),
+                new[] { "TypeDef:<Module>", "TypeDef:C" },
+                new[] { "MethodDef: Void Main()", "MethodDef: Void .ctor()" },
+                new[] { "CompilationRelaxationsAttribute", "RuntimeCompatibilityAttribute", "DebuggableAttribute" }
+                );
+
+            var doc = Path.Combine(dir.Path, "doc.xml");
+            Assert.True(File.Exists(doc));
+
+            using (var reader = new StreamReader(doc))
+            {
+                var content = reader.ReadToEnd();
+                var expectedDoc =
+@"<?xml version=""1.0""?>
+<doc>
+    <assembly>
+        <name>a</name>
+    </assembly>
+    <members>
+        <member name=""M:C.Main"">
+            <summary>Main method</summary>
+        </member>
+    </members>
+</doc>";
+                Assert.Equal(expectedDoc, content.Trim());
+            }
+
+            var output = ProcessUtilities.RunAndGetOutput(exe, startFolder: dir.ToString());
+            Assert.Equal("Hello", output.Trim());
+
+            var refDll = Path.Combine(dir.Path, "ref\\a.dll");
+            Assert.True(File.Exists(refDll));
+
+            VerifyPEMetadata(File.OpenRead(refDll),
+                new[] { "TypeDef:<Module>", "TypeDef:C" },
+                new[] { "MethodDef: Void Main()", "MethodDef: Void .ctor()" },
+                new[] { "CompilationRelaxationsAttribute", "RuntimeCompatibilityAttribute", "DebuggableAttribute" }
+                );
+
+            output = ProcessUtilities.RunAndGetOutput(refDll, startFolder: dir.ToString(), expectedRetCode: -532462766);
+
+            // Clean up temp files
+            CleanupAllGeneratedFiles(dir.Path);
+        }
+
+        private static void VerifyPEMetadata(FileStream peStream, string[] types, string[] methods, string[] attributes)
+        {
+            using (var refPeReader = new PEReader(peStream))
+            {
+                var metadataReader = refPeReader.GetMetadataReader();
+
+                // The types and members that are included needs further refinement.
+                // See issue https://github.com/dotnet/roslyn/issues/17612
+                AssertEx.SetEqual(metadataReader.TypeDefinitions.Select(t => metadataReader.Dump(t)), types);
+
+                AssertEx.SetEqual(metadataReader.MethodDefinitions.Select(t => metadataReader.Dump(t)), methods);
+
+                // ReferenceAssemblyAttribute is missing.
+                // See issue https://github.com/dotnet/roslyn/issues/17612
+                AssertEx.SetEqual(
+                    metadataReader.CustomAttributes.Select(a => metadataReader.GetCustomAttribute(a).Constructor)
+                        .Select(c => metadataReader.GetMemberReference((MemberReferenceHandle)c).Parent)
+                        .Select(p => metadataReader.GetTypeReference((TypeReferenceHandle)p).Name)
+                        .Select(n => metadataReader.GetString(n)),
+                    attributes);
```

Comment:
I wrote those incrementally and found this was easier to debug (I'd copy parts of the LINQ chain into the Immediate window) ;-)